Update MPA expand/collapse functionality

sfc-gh-jgarcia opened the pull request:
## 📚 Context

This PR addresses some issues reported by @jrieke and other users [here](https://www.notion.so/streamlit/Papercut-multipage-handle-b09cfe603cba4c15a307edef15a9d4a6) regarding our existing mechanism to expand/collapse the sidebar navigation.

- What kind of change does this PR introduce?

  - [x] Bugfix
  - [x] Feature
  - [ ] Refactoring
  - [ ] Other, please describe:

## 🧠 Description of Changes

- Removed the up and down carets, and replaced that with a more clear wording;
- Removed the overflow behavior, and instead sliced the number of apps shown when the menu is collapsed;
- Check whether the active page has sidebar widgets on it, and update the collapse/expand state based on that (we want the menu to take up as much space as possible when no widgets are below it);
- Store the expand/collapse preference in `localStorage` so it persists when you refresh the page.

  - [ ] This is a breaking API change
  - [x] This is a visible (user-facing) change

**Revised:**

https://user-images.githubusercontent.com/103376966/216356134-b828325e-aada-4af3-ae52-f4971705d29f.mp4

**Current:**

https://user-images.githubusercontent.com/103376966/216356614-c8c64e90-1f2c-4d6a-81f1-6608d31ebc70.mp4

## 🧪 Testing Done

- [x] Screenshots included
- [x] Added/Updated unit tests
- [x] Added/Updated e2e tests

## 🌐 References

- https://www.notion.so/streamlit/Papercut-multipage-handle-b09cfe603cba4c15a307edef15a9d4a6
- https://www.notion.so/streamlit/Post-launch-feedback-multipage-apps-78795ad17e7e40e49dbfcfcbad47cb4f

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-jgarcia committed:
Remove old e2e snapshots

sfc-gh-jgarcia committed:
Add back missing snapshots

sfc-gh-jgarcia committed:
Update inital state for collapse/expand behavior

sfc-gh-jgarcia committed:
Remove unused imports

mayagbarnes commented:
Some of the snapshots are showing what I assume is unintended behavior. The collapse state for MPA menu (when `View less` is shown) displays consistent with your video, but the expand state (when `# More` displays) shows `-1 More`:
<img width="100" alt="Screenshot 2023-02-10 at 12 48 18 PM" src="https://user-images.githubusercontent.com/63436329/218184778-3c22009f-7790-4a3f-b58b-29b117ecb7b7.png">

See the following snapshots:
* hello-mpa-sidebar-nav-dark.snap.png
* hello-mpa-sidebar-nav.snap.png
* welcome-streamlit-dark.snap.png
* welcome-streamlit.snap.png

Essentially the opposite button is being shown than the one you intend. Taking a quick look, the `expanded` slice of state seems to be backwards.

mayagbarnes commented on frontend/src/components/core/Sidebar/styled-components.ts:
```diff
@@ -65,43 +68,14 @@ export interface StyledSidebarNavItemsProps {
   ({ isExpanded, isOverflowing, hasSidebarElements, theme }) => {
-    const isExpandedMaxHeight = isExpanded ? "75vh" : "33vh"
-    const maxHeight = hasSidebarElements ? isExpandedMaxHeight : "100vh"
+    const isExpandedPaddingBottom = isExpanded ? theme.spacing.threeXS : 0
```
Looks like `isOverflowing` prop is no longer used and can be removed from `StyledSidebarNavItems`

sfc-gh-jgarcia replied:
Thanks! Updated on [6919ec3](https://github.com/streamlit/streamlit/pull/6050/commits/6919ec3e85adc6a8573561634ed007127edb35e1)

mayagbarnes commented on frontend/src/components/core/Sidebar/styled-components.ts:
```diff
@@ -15,14 +15,17 @@
+export interface StyledSidebarNavButtonProps {
+  isExpanded: boolean
+}
+
```
Nit: prefer this interface declaration to be right above its usage in `StyledSidebarNavButton` consistent with organization of the rest of the file

sfc-gh-jgarcia replied:
Got it! Updated in [6919ec3](https://github.com/streamlit/streamlit/pull/6050/commits/6919ec3e85adc6a8573561634ed007127edb35e1)

mayagbarnes commented on frontend/src/components/core/Sidebar/SidebarNav.tsx:
```diff
@@ -76,12 +87,16 @@ const SidebarNav = ({
 
   const toggleExpanded = useCallback(() => {
-    if (!expanded && isOverflowing) {
+    if (!expanded) {
       setExpanded(true)
+      setPagesToShow(-1)
+      localStorage.setItem("navExpanded", "true")
     } else if (expanded) {
```
Agree with CodeQL here, since there are only 2 states for `expanded` boolean, an if/else would be sufficient:
```javascript
if (!expanded) {
      setExpanded(true)
      setPagesToShow(-1)
      localStorage.setItem("navExpanded", "true")
    } else {
      setExpanded(false)
      setPagesToShow(6)
      localStorage.setItem("navExpanded", "false")
    }
  }, [expanded])
```

sfc-gh-jgarcia replied:
Makes sense! Updated in [6919ec3](https://github.com/streamlit/streamlit/pull/6050/commits/6919ec3e85adc6a8573561634ed007127edb35e1)

mayagbarnes commented on frontend/src/components/core/Sidebar/SidebarNav.test.tsx:
```diff
@@ -17,12 +17,10 @@
 
-import Icon from "src/components/shared/Icon"
 import { useIsOverflowing } from "src/lib/Hooks"
 import { mount, shallow } from "src/lib/test_util"
```
Quick note: it is our goal on the Open Source team to transition away from using Enzyme so that we can upgrade to React v18, which does not support the Enzyme testing framework. This does not impact usage of Jest.

We have started asking the following for PR approval:
* All new FE unit tests be written using `react-testing-library` (RTL) instead -> removing usage of the `mount` and `shallow` functions, `.find`, etc. 
* If you are editing unit tests, rewrite the enzyme tests to RTL

I realize this might be an aggressive ask in the case of this PR and depending on your familiarity with RTL/Enzyme, so happy to make the exception, but want to take the opportunity to share some resources for future reference:
* [RTL Docs Migration Example](https://testing-library.com/docs/react-testing-library/migrate-from-enzyme/#basic-enzyme-to-react-testing-library-migration-examples)
* Some conversions shown in [this PR](https://github.com/streamlit/streamlit/pull/6084/files)

sfc-gh-jgarcia replied:
Got it, thanks for letting me know, and thanks for suggesting the exception ❤️ 

We can talk a bit more tomorrow during our meeting and decide whether we might do it for this PR or not, if that works for you!

mayagbarnes commented on frontend/src/components/core/Sidebar/SidebarNav.test.tsx:
```diff
@@ -215,27 +214,20 @@ describe("SidebarNav", () => {
+        .find(StyledSidebarNavButton)
+        .props()
+    ).toHaveProperty("isExpanded", false)
   })
```
I think this test could be improved by also checking the text content of the button (`# More`) in the same way the expanded test below does.

Also, I'd like to see a separate test added to check that the number of additional pages show in the `# More` button is calculated correctly in a couple scenarios to avoid strange behavior like the `-1 More` button

sfc-gh-jgarcia replied:
Fixed on [e225815](https://github.com/streamlit/streamlit/pull/6050/commits/e225815d0f1ae8fdcd5abab5a3e32bc99b56eb0b), and added the extra tests on [14a502d](https://github.com/streamlit/streamlit/pull/6050/commits/14a502de2931656e5c67bc48f74ef70b4130e182)

mayagbarnes commented on frontend/src/components/core/Sidebar/SidebarNav.tsx:
```diff
@@ -57,7 +57,18 @@ const SidebarNav = ({
 
-  const [expanded, setExpanded] = useState(false)
+  // Check localStorage to see if the user has a preference set
+  const isLocalStorageSet = localStorage.getItem("navExpanded") !== null
```
This [recently merged PR](https://github.com/streamlit/streamlit/pull/6100) has implications for usage of `localStorage` -> essentially the need to check its availability as a precursor.

sfc-gh-jgarcia replied:
Oh, I see! Thanks for letting me know, updated on [8fef855](https://github.com/streamlit/streamlit/pull/6050/commits/8fef8554a2d7db7e0d1eef487f8aea15e2b50efc)

mayagbarnes commented on frontend/src/components/core/Sidebar/SidebarNav.tsx:
```diff
@@ -57,7 +57,18 @@ const SidebarNav = ({
+    isLocalStorageSet && localStorage.getItem("navExpanded") === "true"
+  const pageLimit = shouldNavExpand === true ? -1 : 6
+
+  const [expanded, setExpanded] = useState(
+    hasSidebarElements === true ? false : shouldNavExpand
+  )
+  const [pagesToShow, setPagesToShow] = useState(
+    hasSidebarElements === true ? 6 : pageLimit
+  )
```
I think it could be helpful to add some explanatory comments to the default state values here (ex: If there are sidebar elements, we want to render a collapsed MPA page nav. If not, we defer to the localStorage preference)

mayagbarnes commented on frontend/src/components/core/Sidebar/SidebarNav.tsx:
```diff
@@ -57,7 +57,18 @@ const SidebarNav = ({
+  const isLocalStorageSet = localStorage.getItem("navExpanded") !== null
+  const shouldNavExpand =
+    isLocalStorageSet && localStorage.getItem("navExpanded") === "true"
+  const pageLimit = shouldNavExpand === true ? -1 : 6
```
It seems weird to me to have a pageLimit be `-1`... could you explain why this is the case versus using 0 for example?
Also same thoughts for `pagesToShow` being set to -1

sfc-gh-jgarcia replied:
> could you explain why this is the case versus using 0 for example?

Yeah! `0` wouldn't work because we're using this variable to slice the array depending on the conditions above, and if we set it to zero, then you'll end up with something like: `appPages.slice(0, 0)`, which might not work as expected.

Your comment though, sparked a better solution which is using `appPages.length` as the upper bound in those cases: I've fixed it on [eb2737f](https://github.com/streamlit/streamlit/pull/6050/commits/eb2737fe42c76d673771f7239ed34e248614c74e), let me know if you think that's more elegant!

mayagbarnes reviewed (changes requested):
If its okay, I'd like to discuss this PR to make sure I understand the intended behavior/goals to provide a more thoughtful review of the core logic in `SidebarNav.tsx`.

sfc-gh-jgarcia committed:
Remove isOverflowing prop and move interface declaration

sfc-gh-jgarcia committed:
Update collapsed button test

sfc-gh-jgarcia committed:
Merge develop

sfc-gh-jgarcia committed:
Remove unused import

sfc-gh-jgarcia committed:
Change -1 for appPages.length in pagesToShow

sfc-gh-jgarcia committed:
Fix on expand/collapse functionality when you have less than 6 pages

sfc-gh-jgarcia committed:
Add more tests

sfc-gh-jgarcia committed:
Use localStorageAvailable before grabbing its data

sfc-gh-jgarcia committed:
Regenerate snapshots

sfc-gh-jgarcia commented:
@mayagbarnes Thanks a lot for this thorough review! I've addressed all your suggestions (except for re-writing the tests), happy to meet tomorrow and talk about these topics.

sfc-gh-jgarcia committed:
Add back removed snapshots

sfc-gh-jgarcia committed:
Move collapse/expand logic to higher components

sfc-gh-jgarcia committed:
Linting fixes

sfc-gh-jgarcia committed:
Add condition to show/hide the collapse buttons

sfc-gh-jgarcia committed:
Move the hide sidebar logic to Sidebar instead of App

sfc-gh-jgarcia committed:
Updates to nav separator

sfc-gh-jgarcia committed:
Update unit tests

sfc-gh-jgarcia committed:
Add new e2e tests

sfc-gh-jgarcia committed:
Update tests

sfc-gh-jgarcia committed:
Add new snapshots

sfc-gh-jgarcia committed:
Fix on expand/collapse button padding

sfc-gh-jgarcia committed:
Check for change in elements in Sidebar component to prevent flickering

sfc-gh-jgarcia commented:
Hey @mayagbarnes! Did a bunch of work on this PR, per our conversation last week. Tried to fix all the issues you brought up, added the extra e2e/unit tests, and tried to clean up and make the code a bit easier to understand.

I also discussed with @tvst the different factors that change the nav's behavior (# of pages, widgets/no-widgets, localStorage preference setup, etc.) and we agreed we should try to mimic the existing MPA behavior, but change the way it gets triggered (meaning, replace the arrow, the overflowing and the scrolling capabilities, and use the `More/Less` buttons instead). Hopefully this version addresses that!

I've stared at this code for way too long now, haha, and I'm pretty sure there are still nitpicks, ways to improve the code, make it more readable and easier to follow, so let me know if you find anything else that could be improved!

As per refactoring the unit tests, I still don't feel comfortable enough with the new testing framework to do so, specially for a rather complex set of test suites like these for the sidebar, so I'd appreciate skipping that this time, if possible. Totally fine if not, let me know and I'll try to carve out some time to work on that.

Thanks a lot for all your help thus far!

mayagbarnes commented on frontend/src/components/core/Sidebar/styled-components.ts:
```diff
@@ -59,96 +58,34 @@ export const StyledSidebarNavContainer = styled.div(({ theme }) => ({
 export interface StyledSidebarNavItemsProps {
   isExpanded: boolean
-  isOverflowing: boolean
   hasSidebarElements: boolean
```
Nit: Looks like neither of the special props remaining are used for styling the `StyledSidebarNavItems`  and the entire interface can be removed

mayagbarnes commented on frontend/src/components/core/Sidebar/styled-components.ts:
```diff
@@ -59,96 +58,34 @@ export const StyledSidebarNavContainer = styled.div(({ theme }) => ({
 )
 
 export interface StyledSidebarNavSeparatorContainerProps {
   isExpanded: boolean
-  isOverflowing: boolean
 }
```
Nit: Same here, looks like the special props are no longer used for styling the `StyledSidebarNavSeparatorContainer`  and the entire interface can be removed.

mayagbarnes commented on frontend/src/components/core/Sidebar/styled-components.ts:
```diff
@@ -318,3 +255,26 @@ export const StyledResizeHandle = styled.div(({ theme }) => ({
   },
 }))
+
+export interface StyledSidebarNavButtonProps {
+  isExpanded: boolean
+}
```
Nit: Same, looks like `isExpanded` is no longer used for styling the `StyledSidebarNavButton` and the entire interface can be removed.

sfc-gh-jgarcia replied:
@mayagbarnes Yeah, I tried removing them, but it leads to errors in the `.tsx` component (see example below):

<img width="842" alt="Screenshot 2023-02-23 at 11 27 19 AM" src="https://user-images.githubusercontent.com/103376966/220936190-f7d1b5db-1b38-4c18-a325-907a9afd24cf.png">

Should we keep them as they are? Or move them to the Typescript component maybe?

mayagbarnes replied:
This error is expected and is just a reminder to remove the special props you no longer need/ defined as being allowed from being passed to the `StyledSidebarNavItems` component.

**FWIW addtl context:**
Previously we explicitly defined an interface `StyledSidebarNavItemsProps` which specified that the `StyledSidebarNavItems` component has two mandatory props extending the typical allowed props - `isExpanded` & `hasSidebarElements`. We define these interfaces when we need the styled components defined in any `styled-components.ts` file to be aware of component information for styling purposes. Thus the `isExpanded` & `hasSidebarElements` props were passed through the `StyledSidebarNavItems` component in the relevant `.tsx` file to provide this info to the styled component. 

If the styled component does not need any additional information (ex: whether the sidebar nav is expanded) other than the theme for styling, then a specially defined interface is not necessary, and special props passed to the `StyledSidebarNavItems` component are not necessary/will cause the error you encountered above.

**TLDR Part 1: Your styled-components.ts file should look like:**
```javascript
// Remove this interface defining special props passed to the styled component
// export interface StyledSidebarNavItemsProps {
//   isExpanded: boolean
//   hasSidebarElements: boolean
// }

// Eliminate the special props reference
export const StyledSidebarNavItems = styled.ul(
  ({ theme }) => {
    return {
      listStyle: "none",
      margin: 0,
      paddingTop: theme.sizes.sidebarTopSpace,
    }
  }
)
```

**TLDR Part 2: Your SidebarNav.tsx file should look like:**
```javascript
<StyledSidebarNavContainer data-testid="stSidebarNav">
     // Eliminate the special props being passed to the styled component
      <StyledSidebarNavItems
        ref={navItemsRef}
      >
// ... rest here
```

sfc-gh-jgarcia replied:
Got it, thanks for the explanation! I've fixed them all in [b61a5c3](https://github.com/streamlit/streamlit/pull/6050/commits/b61a5c34b3a4c1389b3a3d61f0bd4cb89516c827) and updated tests in [29eec8f](https://github.com/streamlit/streamlit/pull/6050/commits/29eec8f27e97e81dda1642b9741ee85875e3a28f)

mayagbarnes commented:
Hey @sfc-gh-jgarcia! Definitely looks like you put in lots of hard work, appreciate it 👍🏼 
Will try to take a deeper look in the next several days, and don't worry about refactoring the unit tests to the new framework for this PR.

sfc-gh-jgarcia committed:
Remove unused interfaces

sfc-gh-jgarcia committed:
Update tests without using isExpanded prop

mayagbarnes commented on frontend/src/components/core/Sidebar/SidebarNav.tsx:
```diff
@@ -47,121 +54,120 @@ export interface Props {
-    } else if (expanded) {
-      setExpanded(false)
+  useEffect(() => {
+    if (collapseNav !== !expanded) {
```
Nit: Prefer to avoid a double negative for readability's sake, perhaps `collapseNav === expanded`?

sfc-gh-jgarcia replied:
Makes sense! Fixed on [773fa1b](https://github.com/streamlit/streamlit/pull/6050/commits/773fa1bcf187fc49d7310c3fd17980db2afdb448)

mayagbarnes commented on frontend/src/components/core/Sidebar/Sidebar.tsx:
```diff
@@ -196,7 +249,12 @@ class Sidebar extends PureComponent<SidebarProps, State> {
   // BUG(tvst): X button should have same color as hamburger.
   // BUG(tvst): X and > buttons should have same margins as hamburger.
```
Hoping to get your feedback on these - as far as I can tell the X and > sidebar buttons have same color & margins as the hamburger menu button, can we remove these TODO items?

sfc-gh-jgarcia replied:
Yeah, that's correct! We fix that a while back on a different papercut. Removed the comments on [773fa1b](https://github.com/streamlit/streamlit/pull/6050/commits/773fa1bcf187fc49d7310c3fd17980db2afdb448)

sfc-gh-jgarcia committed:
Merge branch 'develop' into update-mpa-handle

sfc-gh-jgarcia committed:
Fix on double negative condition and remove old bug reports

mayagbarnes commented on frontend/src/components/core/Sidebar/SidebarNav.tsx:
```diff
@@ -32,13 +37,15 @@ import {
   currentPageScriptHash: string
   hasSidebarElements: boolean
+  collapseNav: boolean
   hideParentScrollbar: (newValue: boolean) => void
```
Looks like we removed the logic for overflow, mouseOver/mouseOut, etc. so just want to confirm - there aren't any cases where we will be using a scrollbar in the MPA nav right?

If we have removed the scrolling from the MPA nav, we should remove the `hideParentScrollbar` prop from `SidebarNav.tsx` and its associated tests.

sfc-gh-jgarcia replied:
Upon reviewing with the product team, we decided we'd like to keep the scrollbar on the MPA _only_ when:
* You have too many pages;
* Nav is expanded;
* There are widgets on the sidebar.

Then and only then, the sidebar will go up to 75% of the height, and allow scrolling. See video of the intended behavior below 👇🏻

https://user-images.githubusercontent.com/103376966/223125521-dbdd97bc-d1d6-4179-a4a7-4879a9a6bde7.mp4

I've reverted this functionality on [41abaf3](https://github.com/streamlit/streamlit/pull/6050/commits/41abaf37e1cadb57982ad4a4c32d1c8faa1db516) and added back the relevant tests. Let me know what you think, and if further adjustments are needed here!

mayagbarnes commented on frontend/src/components/core/Sidebar/Sidebar.tsx:
```diff
@@ -51,6 +51,7 @@ export interface SidebarProps {
+  collapsedSidebarNav: boolean
   sidebarWidth: string
   lastInnerWidth: number
 
```
Same for this file -> If we have removed the scrolling from the MPA nav, we should remove the `hideScrollbar` slice of state, the `hideScrollbar` function, the `hideParentScrollbar` prop passed to `SidebarNav.tsx` and all associated tests.

sfc-gh-jgarcia replied:
Per my [comment above](https://github.com/streamlit/streamlit/pull/6050#discussion_r1126425684), we'd like to keep the scrolling behavior in one particular use case, so we'll need to keep these

mayagbarnes commented on frontend/src/components/core/Sidebar/Sidebar.tsx:
```diff
@@ -121,6 +157,23 @@ class Sidebar extends PureComponent<SidebarProps, State> {
+  static shouldNavCollapse(
+    pageQuantity: number,
+    hasElements: boolean,
+    collapsedPreference: boolean
```
I think the `collapsedPreference` param should be named `expandedPreference` based on how this function's logic and use. In both calls to `shouldNavCollapse` this input is `cachedSidebarNavExpandedPreference` (expand true, collapse false).

sfc-gh-jgarcia replied:
Makes sense! Fixed on [91e9a74](https://github.com/streamlit/streamlit/pull/6050/commits/91e9a7441e3a44d625374a6537b965b0ea952228)

mayagbarnes commented on frontend/src/components/core/Sidebar/Sidebar.tsx:
```diff
@@ -101,6 +128,15 @@ class Sidebar extends PureComponent<SidebarProps, State> {
       })
     }
+
+    // Change collapse/expand preference if there are widgets on the page
+    this.setState({
+      collapsedSidebarNav: Sidebar.shouldNavCollapse(
+        this.props.appPages.length,
+        this.props.hasElements,
+        cachedSidebarNavExpandedPreference
+      ),
+    })
```
These lines should be removed - we shouldn't have an unconditional call to `setState` in `componentDidUpdate`, but also this causes an unnecessary re-render of the component. Changes in props would already trigger a re-render, and `componentDidUpdate` is called thereafter.
Visually, it results in jumpy behavior, but is also a performance hit.

sfc-gh-jgarcia replied:
I 100% agree with the unconditional call to `setState`, but for some reason, when I added the `this.props.hasElements !== prevProps.hasElements` condition before the state update, it wouldn't work as expected... Might be related to how I was passing on props before.

Anyway, it does work as expected now, so I added the condition on [c861be6](https://github.com/streamlit/streamlit/pull/6050/commits/c861be6b8969b05eace5c913f422cfe125bab6ab). If I remove the `setState` call entirely though, the navigation won't automatically collapse/expand depending on whether you have/don't have widgets in the sidebar, which is a behavior we'd like to keep. Let me know if you have a better way to address that, and I'll get it fixed!

mayagbarnes commented on frontend/src/components/core/Sidebar/Sidebar.tsx:
```diff
@@ -92,6 +115,10 @@ class Sidebar extends PureComponent<SidebarProps, State> {
     this.mediumBreakpointPx = Sidebar.calculateMaxBreakpoint(
       this.props.theme.breakpoints.md
     )
+
+    const cachedSidebarNavExpandedPreference =
+      Sidebar.getSidebarNavPreference()
+
```
This can then be removed because we are removing the `setState` call below that uses it

sfc-gh-jgarcia replied:
We might still need it, per [my comment above](https://github.com/streamlit/streamlit/pull/6050#discussion_r1121723933), but let me know if you have a better way to address this, and I'll get it fixed!

mayagbarnes commented on frontend/src/components/core/Sidebar/Sidebar.test.tsx:
```diff
@@ -137,6 +142,33 @@ describe("Sidebar Component", () => {
+
+    expect(wrapper.find(SidebarNav).exists()).toBe(true)
+  })
+
```
Since the logic for `shouldNavCollapse` function and resulting `collapseNav` prop passed to the SidebarNav lives in the Sidebar component, I think it would be good to test the expected outcomes. For example:
1. pageQuantity < 7 results in `collapseNav` prop  false
2. no sidebar elements results in false
3. user preference expanded (probably have to mock it) results in false
4. pageQuantity of 7, has sidebar elements, user pref collapse results in true

sfc-gh-jgarcia replied:
Added them in [98d8011](https://github.com/streamlit/streamlit/pull/6050/commits/98d8011883089ae54f686b6fc66f77c88fd55c0b)

mayagbarnes commented on frontend/src/components/core/Sidebar/SidebarNav.test.tsx:
```diff
@@ -387,16 +383,6 @@ describe("SidebarNav", () => {
     expect(props.hideParentScrollbar).not.toHaveBeenCalled()
   })
```
This test should be removed - no longer any `mouseOver` behavior

sfc-gh-jgarcia replied:
As mentioned on [my comment above](https://github.com/streamlit/streamlit/pull/6050#discussion_r1126425684), we decided to bring this functionality back for certain use cases, so we might want to keep the test

mayagbarnes commented on frontend/src/components/core/Sidebar/SidebarNav.test.tsx:
```diff
@@ -215,129 +201,148 @@ describe("SidebarNav", () => {
-  it("changes cursor to pointer above separator when overflowing", () => {
-    mockUseIsOverflowing.mockReturnValueOnce(true)
-    // Need mount > shallow here so that toHaveStyleRule can be used.
+  it("is collapsed by default if there are widgets on the sidebar", () => {
```
Since this condition is no longer the case (collapse/expanded default passed via `collapseNav` prop) - it would be great to convert this to confirm that `collapseNav: true` results in 6 instances of `StyledSidebarNavLinkContainer` (or something along those lines), as well as the related reverse scenario, where `collapseNav: false` results in 8 instances of `StyledSidebarNavLinkContainer`

sfc-gh-jgarcia replied:
Added those tests in [90ac45e](https://github.com/streamlit/streamlit/pull/6050/commits/90ac45eeb4c394e4a999775f7bc1340b96c7d4bd)

mayagbarnes commented on frontend/src/components/core/Sidebar/SidebarNav.test.tsx:
```diff
@@ -215,129 +201,148 @@ describe("SidebarNav", () => {
   })
 
-  it("does not render an icon when not expanded and not overflowing", () => {
+  it("doesn't render the StyledSidebarNavButton if pages < 6", () => {
```
Nit: if pages <= 6

sfc-gh-jgarcia replied:
Fixed on [e9d8489](https://github.com/streamlit/streamlit/pull/6050/commits/e9d84892fa16232119bce232c7ecda537b188d09)

sfc-gh-jgarcia committed:
Fix on test definition

sfc-gh-jgarcia committed:
Rename collapsedPreference param

sfc-gh-jgarcia committed:
Add condition to setState call

sfc-gh-jgarcia committed:
Revert and use isOverflowing behavior for certain use cases

sfc-gh-jgarcia committed:
Add sidebar unit tests

sfc-gh-jgarcia committed:
Update sidebar nav tests

sfc-gh-jgarcia committed:
Regenerate e2e tests

sfc-gh-jgarcia committed:
Add back e2e specs
